core/fed_avg_optimizer.py

```diff
@@ -0,0 +1,157 @@
+import logging
+
+from core.utils.enums import ActionableEventTypes, RawEventTypes
+from core.utils.enums import JobTypes, callback_handler
+from core.utils.dmljob import deserialize_job
+from core.utils.keras import serialize_weights
+from core.utils.dmlresult import DMLResult
+
+
+logging.basicConfig(level=logging.DEBUG,
+	format='[FedAvgOpt] %(asctime)s %(levelname)s %(message)s')
+
+
+class FederatedAveragingOptimizer(object):
+	"""
+	Federated Averaging Optimizer
+
+	Implements the Federated Averaging algorithm (with some tweaks)
+	and acts as a "brain" for the Unix Service
+
+	Specifically, upon hearing a new gradient, it will schedule
+	a running average and increment listen_iterations until it exceeds
+	listen_bound, at which point it will start retraining
+
+	Breaks down info into Level_1 and Level_2 callbacks. 
+	Level_1 Callbacks determine which Level_2 Callback to call,
+	and do some Level_2_ -specific action to make the payload
+	ready for the Level_2 Callback.
+
+	"""
+	def __init__(self, content):
+		"""
+		Kwargs is a dictionary containing all the info needed to initialize the Optimizer.
+		This includes any number of properties such as thresholds, etc.
+		"""
+		logging.info("Setting up Optimizer")
+		kwargs = content.get('optimizer_params')
+		serialized_job = content.get('serialized_job')
+		self.job = deserialize_job(serialized_job) # The Optimizer knows its job
+		self.listen_iterations = kwargs.get('listen_iterations', 0)
+		self.listen_bound = kwargs.get('listen_bound', 1)
+		self.LEVEL_1_CALLBACKS = {
+			RawEventTypes.JOB_DONE.name: self._handle_job_done,
+			RawEventTypes.NEW_INFO.name: self._handle_new_info,
+			RawEventTypes.NOTHING.name: self._do_nothing,
```

Comment:
Why would the Communication Manager ask the Optimizer what to do with "NOTHING"?

Reply:
sure? you initially were the one who wanted to have these default "NOTHING" values so idrc, removed